Product Gallery Thumbnails: remove settings from container block

kmanijak opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes # .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to the Editor > Single Product template.
2. Replace the Product Image Gallery block with the Product Gallery (beta) block.
3. Click on the Product Gallery (Beta) block
4. **Expected:** Make sure there's **no** "Thumbnails" settings like these:
<img width="283" alt="image" src="https://github.com/user-attachments/assets/c869590a-55e8-4434-926a-8b4527c2248c" />

6. Click on the Thumbnails block
7. **Expected:** Make sure there's **no** "Thumbnails" settings like those below:
8. Try:
    - changing position of Thumbnails by changing blocks order (see video below)
    - changing Gallery Area block layout (Row -> Stack and opposite)
    - changing the number of thumbnails
    - save and go to frontend in various moments
12. Expected: Make sure "Thumbnails" are displayed either horizontally (above and below large image) or vertically (on sides), adjust to different number of thumbnails and display similarly on the frontend

Note: This PR doesn't cover ALL of the cases of Gallery Area layout settings. Block may end up in some weird state in edge cases but should behave correctly with the default Row or Stack layouts.

Before | After
-- | --
https://github.com/user-attachments/assets/3fcfe2ce-5afd-4f3c-b0bc-399618523216 | https://github.com/user-attachments/assets/8daf35a2-4c6f-4995-8296-00dc9b9f2a12

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

kmanijak committed:
Remove thumbnails settings from container block

kmanijak committed:
Migrate Thumbnails from context usage to local attributes

kmanijak committed:
Cleanup types

kmanijak committed:
Adjust frontend

kmanijak committed:
Fix the frontend gallery

kmanijak committed:
Fix types

kmanijak committed:
Fix types

kmanijak committed:
Remove provide context

kmanijak committed:
Remove occurrances of productGalleryClientId

kmanijak committed:
Update block references and block manifest

kmanijak committed:
Rename thumbnailsNumberOfThumbnails to numberOfThumbnails

kmanijak committed:
Remove attribute from gallery template

kmanijak committed:
Remove thumbnails editor styles

kmanijak committed:
Simplify Thumbnails edit file

kmanijak committed:
Remove thumbnails position manipulation from parent

kmanijak committed:
Remove utils manipulating positions

kmanijak committed:
WIP: Simplify styles

kmanijak committed:
Some stylingimprovements

kmanijak committed:
Merge branch 'trunk' into add/54512

kmanijak committed:
Update block manifest

kmanijak committed:
Bring back some of the old styles

kmanijak committed:
Fix styles

kmanijak committed:
Remove Thumbnails position attribute and corresponding code

kmanijak committed:
Remove leftover exports

kmanijak committed:
Continue adjusting styles

kmanijak committed:
Fix frontend by attaching new classes

kmanijak committed:
Merge branch 'trunk' into add/54512

kmanijak committed:
Remove unnecessary variables and fix merge

kmanijak committed:
Add changelog

kmanijak committed:
Fix lint

kmanijak committed:
Remove tests related to positioning settings

kmanijak committed:
Merge branch 'trunk' into add/54512

kmanijak committed:
Fix E2E

kmanijak committed:
Update blocks references and manifest

kmanijak committed:
Fix test

tjcafferkey commented:
I don't know if this is something to fix in this PR but in this scenario of settings (thumbnails left, 6 thumbnails) the spacing above/below the thumbnails isn't symmetrical it seems. 

![Screenshot 2025-01-30 at 09 57 37](https://github.com/user-attachments/assets/1043b37b-0a52-4353-af1b-bfa0e7bcffea)

tjcafferkey commented on plugins/woocommerce-blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/edit.tsx:
```diff
@@ -2,66 +2,54 @@
+	setAttributes,
+}: BlockEditProps< ProductGalleryThumbnailsBlockAttributes > ) => {
 	const blockProps = useBlockProps( {
 		className: clsx(
```
We can improve this and remove the `clsx` dependency.
```suggestion
		className: `wc-block-product-gallery-thumbnails wc-block-product-gallery-thumbnails--number-of-thumbnails-${ attributes.numberOfThumbnails }`,
```

tjcafferkey commented on plugins/woocommerce-blocks/assets/js/blocks/product-gallery/style.scss:
```diff
@@ -1,17 +1,9 @@
-$outside-image-max-width: calc(100% - (2 * $outside-image-offset));
 $thumbnails-gap: 15px;
-$default-number-of-thumbnails: 3;
+$thumbnails-border-width: 1px;
```
Nit pick, but this variable is only used once so I'm not sure it needs to be a variable at all. I'm fine either way.

tjcafferkey commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductGalleryThumbnails.php:
```diff
@@ -120,7 +120,7 @@ protected function should_display_view_all( $thumbnails_count, $product_gallery_
 	 * @return string Rendered block type output.
 	 */
 	protected function render( $attributes, $content, $block ) {
```
Do you think we should early return an empty string if there are no gallery thumbnails (or one image)? E.g.

```
if ( ! $product_gallery_images || count( $product_gallery_images ) <= 1 ) {
    return '';
}
```

kmanijak replied:
Good catch!

tjcafferkey reviewed (approved):
Looks good and tests well! Ive just left a few comments but I wouldn't consider them blocking.

kmanijak committed:
Merge branch 'trunk' into add/54512

kmanijak committed:
Address code review

kmanijak committed:
Update blocks manifest

kmanijak committed:
Fix typo

kmanijak committed:
Limit the number of thumbnails if there's not enough images

kmanijak committed:
Fix PHP error

kmanijak committed:
Fix lint

kmanijak committed:
Merge branch 'trunk' into add/54512

kmanijak commented:
> I don't know if this is something to fix in this PR but in this scenario of settings (thumbnails left, 6 thumbnails) the spacing above/below the thumbnails isn't symmetrical it seems.
> Edit: Ah, I see that its because I have 5 thumbnails but my setting says show 6. I assume the code is anticipating 6 thumbnails whereas we should change this assumption to be "maximum" of 6 with the possibility of less.

Good suggestion! I implemented the change in https://github.com/woocommerce/woocommerce/pull/54799/commits/a7de60ff1bd2158c50bb7296226a2828e598d4af!

I also addressed other comments, let me know @tjcafferkey if you want to take a look again or I'm good to merge.

BTW in scope of this PR I also enabled browsing all images in "small" gallery, so basically removed the artificial limit by "number of thumbnails". I think it improves UX and makes Pager showing the correct information. Example: number of thumbnails is 3 but I'm allowed to see beyond that (6/8):

<img width="551" alt="image" src="https://github.com/user-attachments/assets/d67a4bbf-12fb-4ce9-b4d3-d93c4c703557" />

tjcafferkey commented:
@kmanijak I think we're good to merge, thanks for fixing all that! I am just going to note something for discussion that we can decide to do in a follow-up PR if we like.

One thing that I keep noticing is that the alignment of thumbnails look 'off' even though technically they are not. But it is because in both the editor and the frontend the vertical thumbnails (left/right) don't sit flush with the top of the image gallery because they take the Pager block into account when calculating their position.

Do you think these would look better if the top thumbnail was aligned with the top of the Large Image block in both editor and frontend?

![Screenshot 2025-01-31 at 11 58 21](https://github.com/user-attachments/assets/ba1c378e-bb5c-46e7-9945-55160c33b68c)

![Screenshot 2025-01-31 at 12 00 41](https://github.com/user-attachments/assets/0f1aed21-2d09-490d-b152-4fc2c98383bf)

kmanijak commented:
> One thing that I keep noticing is that the alignment of thumbnails look 'off' even though technically they are not. But it is because in both the editor and the frontend the vertical thumbnails (left/right) don't sit flush with the top of the image gallery because they take the Pager block into account when calculating their position.
> Do you think these would look better if the top thumbnail was aligned with the top of the Large Image block in both editor and frontend?

This can be controlled by user in the Gallery Area container settings:
- by default when inserting the block I set vertical alignment to top
- after switching Row to Stack and back to Row, alignment is set to center by default (it's core's Group block behaviour). It can be brought back to top alignment manually:

https://github.com/user-attachments/assets/f34bd0bf-a8b6-4821-9eb2-a4d55ec71207

I think that aligns with overall approach to set good defaults but leave it for manual customizations later. WDYT?

tjcafferkey commented:
>I think that aligns with overall approach to set good defaults but leave it for manual customizations later. WDYT?

Makes sense, thanks for the explanation 👍🏻 lets merge! 🚀
